packages/mobx-state-tree/src/middlewares/on-action.ts

```diff
@@ -81,16 +81,34 @@ function serializeTheUnserializable(baseType: string) {
 export function applyAction(
     target: IStateTreeNode,
     actions: ISerializedActionCall | ISerializedActionCall[]
-): void {
+): Promise<any[]> {
     // check all arguments
     if (process.env.NODE_ENV !== "production") {
         if (!isStateTreeNode(target))
             fail("expected first argument to be a mobx-state-tree node, got " + target + " instead")
         if (typeof actions !== "object")
             fail("expected second argument to be an object or array, got " + actions + " instead")
     }
-    runInAction(() => {
-        asArray(actions).forEach(action => baseApplyAction(target, action))
+
+    return new Promise((res, rej) => {
+        runInAction(() => {
```

Comment:
I am wondering whether async could be opt-in here; e.g. run synchronosly until there is an action that returns a promise, and then start chaining them together?

Reply:
@mweststrate Did you mean:
1. if each action that passed in `applyAction` is synchronous, each action should be executed synchronously and finally return one promise.
2. if any action that passed in `applyAction` is asynchronous, just wait for those async actions fulfilled, then chain their results and the results from sync actions that were synchronously executed. After that, returns one promise.

The point is that execute synchronous actions at the first tick.

Is this what you want?